File: Java/1481.LeastNumberUniqueIntegersAfterKRemovals/Solution.java
/* Greedy: Time:O(nlogn), Space:O(n)
 * 1. Construct num2Count and sort the counts using increasing order
 * 2. Subtract k with counts.get(i) until k <= 0
 */     

import java.util.*; // Stack

public class Solution {
    public int findLeastNumOfUniqueInts(int[] arr, int k) {
        Map<Integer, Integer> num2Count = new HashMap<>();
        for(int num: arr){
            num2Count.put(num, num2Count.getOrDefault(num, 0) + 1);
        }
        List<Integer> counts = new ArrayList<>(num2Count.values());
        Collections.sort(counts);
        for(int i = 0; i < counts.size(); ++i){
            if(k == 0){
                return counts.size() - i;
            }else if(k < 0){
                return counts.size() - i + 1;
            }else{
                k -= counts.get(i);
            }
        }
        return (k >= 0)? 0: 1;
    } 
    
    public static void main(String[] args){
        Solution sol = new Solution();
        int[] arr = {5, 5, 4};
        int k = 1;
        System.out.println("arr:" + Arrays.toString(arr));
        System.out.println("k:" + k);
        System.out.println("least number of uniques:" + sol.findLeastNumOfUniqueInts(arr, k));
    }
}

File: Go/1481.LeastNumberUniqueIntegersAfterKRemovals/Solution.go
/* Greedy: Time:O(nlogn), Space:O(n)
 * 1. Construct num2Count and sort the counts using increasing order
 * 2. Subtract k with counts.get(i) until k <= 0
 */

package main

import ("fmt"
        "sort")

func findLeastNumOfUniqueInts(arr []int, k int) int {
    var num2Count map[int]int = make(map[int]int)
    for _, num := range arr{
        num2Count[num]++
    }
    var counts []int
    for _, val := range num2Count{
        counts = append(counts, val)
    }
    sort.Ints(counts)
    for idx, count := range counts {
        if(count == k) {
            return len(counts) - idx - 1
        }else if(count > k){
            return len(counts) - idx
        }else{
            k = k - count
        }
    }
    return 0
}

func main() {
    var arr = []int{5, 5, 4}
    var k = 1
    fmt.Println("arr:", arr, ", k", k)
    fmt.Println("least number of uniques:", findLeastNumOfUniqueInts(arr, k))
}
